Squeeze edge method

stavros11 opened the pull request:
[Squeeze](https://www.tensorflow.org/api_docs/python/tf/squeeze) edge method (code is exact copy from the internal). I was going to remove this because I do not think it is very useful after all and will save us ~40 lines, but up to you to decide.

Thenerdstation commented on tensornetwork/tensornetwork.py:
```diff
@@ -966,6 +966,44 @@ def contract_parallel(self, edge: Edge) -> Node:
+    """Squeezes a dangling edge with unit size.
+
+    Args:
+      edge: Edge
```
Please improve the documentation

Thenerdstation commented on tensornetwork/tensornetwork.py:
```diff
@@ -966,6 +966,44 @@ def contract_parallel(self, edge: Edge) -> Node:
     return self.contract_between(edge.node1, edge.node2)
 
+  def squeeze(self, edge: Edge) -> Node:
+    """Squeezes a dangling edge with unit size.
```
Also say that this removes the edge from the node.

amilsted replied:
Why restrict this to dangling edges?

Thenerdstation replied:
This basically just removes unit sized dangling edges. Non-dangling unit sized edges are either outer products or no-op contractions.

amilsted replied:
Yes, I know. I think it makes just as much sense to squeeze them away as it does to squeeze dangling edges.

Thenerdstation commented:
Yeah. I'm ok with not including it either (not sure where we would use it). Feel free to close the PR.

amilsted commented:
What currently happens if we do a truncated SVD to split a node and the resulting singular-value dimension is of size 1? I assume we currently still assign an edge to it? I suppose one could use something like squeeze() to remove that (trivial) edge. 

It doesn't seem *necessary* to have this built in, but it's not an operation one can easily do manually, since it requires manipulating the network in many places.

Thenerdstation commented:
I don't see how the SVD is a special case. If there was a unit dangling edge after the SVD, it had to exist before the SVD as well. Unit sized non-dangling edges can just be contracted normally.

stavros11 commented:
Ash probably meant the middle (non-dangling) edge that results from SVD split and did not exist before. This can be unit if there is only one singular value greater than `max_truncation_err`. We can easily extend `squeeze` to contract unit non-dangling edges but I guess the user can call just `contract` on them directly anyway.

amilsted commented:
Ah, but `contract` would not do the same as `squeeze` for a non-dangling edge. At least not in my head :)
It would change the shape of each tensor and remove an edge, not do an outer product.

stavros11 commented:
So I guess the equivalent use would be to call `disconnect` in the non-dangling edge and then squeeze the two new dangling edges. I can add this to squeeze if it is useful.

amilsted commented on tensornetwork/tensornetwork.py:
```diff
@@ -966,28 +966,36 @@ def contract_parallel(self, edge: Edge) -> Node:
-  def squeeze(self, edge: Edge) -> Node:
-    """Squeezes a dangling edge with unit size.
+  def squeeze(self, edge: Edge) -> Union[Node, Tuple[Node]]:
+    """Squeezes an edge with unit size.
```
I think we can come up with a better one-line description, preferably one that doesn't refer to "squeezing". How about "Removes an edge with unit size."?

amilsted commented on tensornetwork/tensornetwork.py:
```diff
@@ -966,28 +966,36 @@ def contract_parallel(self, edge: Edge) -> Node:
+    if not edge.is_dangling():
+      dangling1, dangling2 = self.disconnect(edge)
+      node1 = self.squeeze(dangling1)
+      node2 = self.squeeze(dangling2)
```
Perhaps we should check the dimension of both axes before squeezing anything? Currently, if the axis of node2 (erroneously) has non-unit size, this may modify the network *before* throwing an exception.

Also, note that the dimension may be `None`.

amilsted replied:
The size being `None` should not be an error - this occurs if the size of a dimension is dynamically determined during the execution of a graph and should be allowed. If the axis has dimension `None`, we should instead just let it go through and allow `tf.squeeze` to fail if the dynamic dimension was not 1.

Actually, I don't know if we should try to check the size at all. In `connect()` we do *not* check for compatible sizes, presumably because of the `None` case. Probably we should either check dimensions (where not `None`) in both places, or in neither.

Thenerdstation replied:
We should be checking sizes in both cases.

I think the best we can do is to do static checks outside of `tf.assert_equal` and just let `None`s slip by. Super not ideal, but having slow `Edge` creation is even worse.

amilsted replied:
That would be my preference too. Check using `tensor.shape`. If `tensor.shape[i] is None` we let it pass.

We *could* add a `tf.assert_equal()` op only in case the shapes are undetermined, but I don't think it adds much... Also, one is supposed to explicitly make dependent graph ops depend on the assert op succeeding, which would be horrible to do within `TensorNetwork`.

stavros11 commented:
Thanks for the comments. I changed the size check because comparisons with `get_dimension` might break in graph mode as it returns a `tf.Tensor`. Also added a check of whether the two axes have the same size, however I suspect that this is not necessary as this error should come up during the construction of the network (eg. when one tries to connect the two incompatible axes).

Regarding the `None` case, `tf.squeeze` works well and decides whether to squeeze during run time, when shapes are defined. I am not sure how this can be transferred in our case as it is not a good idea to modify anything in the network before assigning the shapes. So, I raise error for now.

Thenerdstation commented:
> as this error should come up during the construction of the network (eg. when one tries to connect the two incompatible axes).

You would hope, but sadly we don't :disappointed:.

Adding the `tf.assert_equal` call made creating edges incredibly slow. This issue was raised in stefan's repository.

Thenerdstation commented:
I vote we remove this. I'm still not sold that it's ever needed.

stavros11 commented:
From my side you are free to close.
